Post for C/C++ issues on CRAN (#630)
tidyverse/tidyverse.org: content/blog/cran-checks-compiled-code/index.Rmd

---
output: hugodown::hugo_document

slug: cran-checks-compiled-code

> title: "New CRAN Requirements for Packages with C and C++"

Author: May as well be consistent. Thanks!

date: 2023-03-13
author: Andy Teucher
description: >
    A few recent changes in CRAN requirements for packages containing C or C++
    code have caused package developers some headhaches. This post outlines
    the issues and provides solutions the tidyverse team has used to address them.

photo:
  url: https://unsplash.com/photos/dhGFLj3rI0Q
  author: Quino Al

# one of: "deep-dive", "learn", "package", "programming", "roundup", or "other"

> categories: [learn, programming]

Author: Should we try to do this for this post? The icons are here: https://github.com/tidyverse/tidyverse.org/tree/main/themes/hugo-graphite/static/css/images and the categories are defined here: https://github.com/tidyverse/tidyverse.org/blob/72bb2dd7ca0b2f211498a891aa54f55ddcad5014/themes/hugo-graphite/static/css/main-site.css#L835-L856

tags: [cran, c, c++]
---

<!--
TODO:
* [x] Look over / edit the post's title in the yaml
* [x] Edit (or delete) the description; note this appears in the Twitter card
* [x] Pick category and tags (see existing with `hugodown::tidy_show_meta()`)
* [x] Find photo & update yaml metadata
* [x] Create `thumbnail-sq.jpg`; height and width should be equal
* [x] Create `thumbnail-wd.jpg`; width should be >5x height
* [x] `hugodown::use_tidy_thumbnails()`
* [ ] `Add intro sentence, e.g. the standard tagline for the package
* [ ] `usethis::use_tidy_thanks()`
-->

The R package landscape is dynamic, with changes in infrastructure common, especially when CRAN makes changes to their policies and requirements. This is particularly true for packages that include low-level compiled code, requiring developers to be nimble in responding to these changes.

> The tidyverse team at Posit is in the unique situation where we have a concentration of skilled developers whose job it is to create and maintain open source packages. This internal community provides the opportunity to collaborate to develop best practices and discover solutions to problems that arise. When we can, we like to share what we've learned so other developers can benefit.

Reviewer: I'd also suggest 'best practices' -> 'shared practices'

### How to fix it

1.  Edit the DESCRIPTION file and remove `SystemRequirements: C++11`.
2.  Edit `src/Makevars`, `src/Makevars.win`, and `src/Makevars.ucrt` and remove `CXX_STD=CXX11`.

After making these changes, the package should install without trouble on R 3.6 and above. It may not build on R <= 3.5 on systems with very old compilers, though it is likely that the vast majority of users will have a newer version of R and/or have recent enough compilers. If you want to be confident that your package will be installable on R 3.5 and below with old compilers, there are several options; we offer two of the simplest approaches here:

-   You can use a configure script at the top level of the package, and have it add `CXX_STD=CXX11` for R 3.5 and below. An example (unmerged) [pull request to the readxl](https://github.com/tidyverse/readxl/pull/722/files) package demonstrates this approach. You will also need to add `Biarch: true` in your DESCRIPTION file. This appears to be the approach preferred by CRAN.
-   For users with R <= 3.5 on a system with an older compiler, package authors can instruct users to edit their `~/.R/Makevars` file to include this line: `CXX_STD=CXX11`.

> The tidyverse has a [policy of supporting four previous versions](https://www.tidyverse.org/blog/2019/04/r-version-support/) of R. Currently that includes R 3.5, but with the upcoming release of R 4.3 the minimum version we will support is R 3.6. As we won't be supporting R 3.5 in the near future, you should not feel pressured to either unless you have a compelling reason.

Reviewer 2: The date of 4.3.0 is not known yet.

> The tidyverse has a [policy of supporting four previous versions](https://www.tidyverse.org/blog/2019/04/r-version-support/) of R. Currently that includes R 3.5, but with the upcoming release of R 4.3 (which should be this Spring some time) the minimum version we will support is R 3.6. As we won't be supporting R 3.5 in the near future, you should not feel pressured to either unless you have a compelling reason.

Reviewer 3: ```suggestion
The tidyverse has a [policy of supporting four previous versions](https://www.tidyverse.org/blog/2019/04/r-version-support/) of R. Currently that includes R 3.5, but with the upcoming release of R 4.3 (which should be this Spring some time) the minimum version we will support is R 3.6. As we won't be supporting R 3.5 in the near future, you should not feel pressured to either.
```

```
checking compiled code ... WARNING
File 'fs/libs/fs.so':
  Found 'sprintf', possibly from 'sprintf' (C)
    Object: 'file.o'
Compiled code should not call entry points which might 
terminate R nor write to stdout/stderr instead of to the 
console, nor use Fortran I/O nor system RNGs nor [v]sprintf.
See 'Writing portable packages' in the 'Writing R Extensions' manual.
```

According to the [NEWS for R-devel](https://developer.r-project.org/blosxom.cgi/R-devel/NEWS/2022/12/24#n2022-12-24) (which will be R 4.3):

> The use of sprintf and vsprintf from C/C++ has been deprecated in macOS 13 and is a known security risk. `R CMD check` now reports (on all platforms) if their use is found in compiled code: replace by snprintf or vsnprintf respectively.

These are considered to be a security risk because they potentially allow [buffer overflows](https://en.wikipedia.org/wiki/Buffer_overflow) that write more bytes than are available in the output buffer. This is a risk if the text that is being passed to <code>sprintf()</code> comes from an uncontrolled source.

Here is a very simple example:

```{r}
library(cpp11)

cpp_function('

>   int say_height(int height) {

Reviewer 3: Even here I think capturing the `n` is a bit unusual. It's probably best practice, but I think few people do it, and you get to the heart of the problem faster without it.

Author: Fair enough. I was doing it to demonstrate the return value vs the `size` parameter - I've seen places where people capture it and compare it to the expected `size` and then condition on that...

Author: If you're ok with it @hadley I'd rather keep the `n` - I think it helps to demonstrate the slightly different return values of `sprintf()` vs `snprintf()`.

Reviewer 3: I don't have strong feelings about it so go ahead.

Author: Thanks!

  int say_height(int height) {
    // "My height is xxx cm" is 19 characters but we need
    // to add one for the null-terminator
    char out[19 + 1];
    int n;
    n = sprintf(out, "My height is %i cm", height);
    Rprintf(out);
    return n;
  }
'
)

say_height(182)
```

```{r}
#| eval: false
say_height(1824) # This will abort due to buffer overflow
```

### How to fix it

> In most cases, this should be a simple fix: replace <code>sprintf()</code> with `snprintf()` and `vsprintf()` with `vsnprintf()`. These `n` variants take a second parameter that specifies the maximum number of bytes written. If the output is a static buffer, you can use `sizeof()`:

Reviewer 4: Maybe also explain the return value of `snprintf()`? Since it does come up below.

```{r}
cpp_function('
  int say_height_safely(int height) {
    // "My height is xxx cm" is 19 characters but we need 
    // to add one for the null-terminator
    char out[19 + 1];
    int n;

>     n = snprintf(out, sizeof(out), "My height is %i cm\\n", height);

Reviewer 3: That string is length 20, so there's no room for the null terminator?

    n = snprintf(out, sizeof(out), "My height is %i cm", height);
    Rprintf(out);
    if (n >= sizeof(out)) {
       Rprintf("\\nTruncated because input is longer than allowed!\\n");
    }

>     return n;

Reviewer 3: This is probably overkill

Author: Do you mean the "Truncated..." Rprintf output?

Reviewer 3: Yeah

  }
')

say_height_safely(182)
say_height_safely(1824)
```

> If the destination is not a static buffer, the easiest thing to do is pass in the

Reviewer 4: I had cases that were harder to analyze, but I don't have anything productive to say other than "sometimes it's kind of hard to figure this out" 😬

Author: Do you have any examples you could point me to that might help me come up with a better example as per @hadley's [comment](https://github.com/tidyverse/tidyverse.org/pull/630#discussion_r1135822319)?

Reviewer 4: I don't think you can use this directly, but it's an example of a more complicated situation. Here's a patch in the grisu code that is vendored into readr and vroom.

https://github.com/tidyverse/readr/commit/9143718735136c05c52c467b996dff799f77b888

Reviewer 4: And here's where it intersects with code we control:

https://github.com/tidyverse/readr/blob/9558d899b7e8c592be339bfc31ec35593ccf20ca/src/write_delim.cpp#L219

So I guess I could have hardwired a size of 32, but that would have been really opaque on the grisu side.

Reviewer 4: This is the problem with these mandates, because the grisu function should really be refactored to "know" about the size of buffer that's available to it. But with vendored code, you want to make the smallest patch possible.

Author: That's great, thanks! I think it's a great example because you do a nice job explaining the choices you made. I did also just push up a new example, but I think it's pretty contrived. I could link to this instead (or as well)?

Reviewer 4: I'm not sure this readr/grisu thing is a great example because it's still basically static and, ideally, **we** would have captured the rationale for 32 bytes in **our** code. But we didn't.

Maybe @hadley is thinking more of an example where the size of the buffer depends on some data that can't be known in advance. This is cobbled together from reading from SO threads, so if you use it, make sure to get someone else to affirm it's not too dorky!

``` r
library(cpp11)
cpp_function('
  size_t say_hello(std::string name) {
    std::string fmt("Hello, %s!\\n");
    // subtract 2 for %s, add 1 for null terminator
    size_t size = fmt.size() - 2 + name.size() + 1;
    std::string buf;
    buf.resize(size);
    snprintf(&buf[0], size, fmt.c_str(), name.c_str());
    Rprintf(buf.c_str());
    return size;
  }
')
say_hello("Jenny")
#> Hello, Jenny!
#> [1] 15
```

<sup>Created on 2023-03-14 with [reprex v2.0.2.9000](https://reprex.tidyverse.org)</sup>

Reviewer 3: @jennybc I don't love that code either, because if you're using C++, I think you'd generally do something other than `sprintf()`. I can see places where we might use code like that, but it feels a bit artificial.

Maybe we can just omit an example for the case of a dynamic buffer since it will depend so much on the specific scenario.

Reviewer 4: I searched r-lib and tidyverse packages for `snprintf()` and we really don't have any organic examples of a true dynamic buffer.

Reviewer 4: @ateucher If we're going to keep the link and reference to the readr/grisu thing, I'd like the chance to slightly rephrase the text.

Author: I'm not at all opposed to taking out the dynamic buffer example - I don't think the intent is to show people _how_ to figure out the size of something, just to let them know that they need to with this change. It's also at risk of just getting too long...

@jennybc I'll leave it up to you if we keep it in - and definitely please go ahead and edit it if you think it's worth it. Though I think you have a good point about doing some gymnastics so as to not refactor it properly, so maybe it's not the best illustration?

Reviewer 4: Let's just drop it, because it's such a mixed bag.

returns the length of the formatted string, whether or not it has been truncated to match `size`.

> It is a bit trickier if the destination is not a static buffer, so you must set a maximum `size` that you 
> think is reasonable:

Reviewer 3: ```suggestion
It is a bit trickier if the destination is not a static buffer, so you'll have to determine the maximum `size` by carefully thinking about the code:
```

Reviewer 3: The example is better, thanks!

  int say_height_safely(int height) {
    // "My height is xxx cm" is 19 characters but we need 
    // to add one for the null-terminator
    size_t size = 19 + 1; 
    char out[size]; 
    int n; 
    n = snprintf(out, size, "My height is %i cm", height);
    Rprintf(out);
    if (n >= sizeof(out)) {
       Rprintf("\\nTruncated because input is longer than allowed!\\n");
    }
    return n;
  }

> ')

Author: I pushed a new example... don't know if it's too contrived?

### How to fix it

This can be fixed by placing the `void` keyword in the previously empty argument list:

``` c
int myfun(void) {
  ...
};
```

Here is an example where the authors of [Cubist](https://topepo.github.io/Cubist/) applied the [necessary patches](https://github.com/topepo/Cubist/pull/46), and [another one in rlang](https://github.com/r-lib/rlang/pull/1508).

### Vendored code

Function declarations without a prototype are very common, and unfortunately are thus likely to appear in libraries that you include in your package. This may require you to patch that code in your package. The [readxl](https://readxl.tidyverse.org) package includes the [libxls C library](https://github.com/libxls/libxls), which was patched [in readxl here](https://github.com/tidyverse/readxl/commit/afdc9b90cfc2bb1e1c5490c7ba3af5ecfc4a7876) to deal with this issue.

> The ideal solution in cases like this would be to submit patches to the upstream libraries so you don't have to deal with the ongoing maintenance of your local patches, but that is not always possible.

Reviewer 3: Generally, however, you can explain this problem when submitting your package, and as long as you've have notified the upstream maintainer, CRAN will accept your updated package.
